### PR TITLE
Redesign [WIP]

### PR DESCRIPTION
**Pull request by davidrleonard**
**WIP -- DO NOT MERGE**

This branch holds the redesigned px-app-nav component. This is a major revision that includes a complete visual redesign and several breaking API changes.

### What's changed

* Visual redesign that introduces multiple "styles" of navigation, including a horizontal, collapsed, and vertical
* New nav `items` API that does not mutate the items it receives -- the state of the app nav can be set using the two-way binding `selectedPath` and `selectedSubpath` attributes
* New `selectedItem` and `selectedSubitem` read-only attributes notify a reference to the item in the nav `items` array. This allows developers to attach arbitrary metadata to an item and data-bind to that metadata. This is useful to manage misc. app state (like the active page name) and bind it to the nav.
* The component is split into the main `<px-app-nav>`  tag, which will be used by developers to build a nav; and a variety of internal subcomponents that are used to modularize the behaviors and styles of the component (`<px-app-nav-item>`, `<px-app-nav-group>`, `<px-app-nav-subgroup>`).

### What to review

It'd be great if we reviewed using the [Github pull request review](https://help.github.com/articles/about-pull-request-reviews/) feature, which allows everyone to comment inline on the code in Github. Please leave feedback, questions, suggestions, etc. that you think would help to make the component better. **I'd especially like feedback on using the public `px-app-nav` API to supply items, select paths/subpaths, get notifications about selections, and so on. I want to nail down the API really well before release so we are in a good place to grow with it.**

You can find simple, working examples in `examples/`. Feel free to start there if you want to mess with something in your browser.

(Note: this code isn't transpiled yet, so it won't work in ES5-only browsers. I'll work on getting it transpiled so we can test it out in IE/old Safari/etc.)

* [ ] px-app-nav.html -- ready to review
* [ ] px-app-nav.es6.js -- ready to review
* [ ] px-app-nav-behavior-measure-text.html - ready to review
* [ ] sass/** -- ready to review
* [ ] px-app-nav-item.html -- refactoring, review if you'd like
* [ ] px-app-nav-group.html -- refactoring, review if you'd like
* [ ] px-app-nav-subgroup -- refactoring, review if you'd like

### Still left to do

* [ ] refactor/unit test px-app-nav-item
* [ ] refactor/unit test px-app-nav-group
* [ ] refactor/unit test px-app-nav-subgroup
* [ ] update docs page to include more API props

### CODE REVIEW

File: px-app-nav-behavior-measure-text.es6.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+     * so it can retrieved later without rebuilding it.
+     *
+     * @param  {String} fontName   A valid [CSS font-family list](https://developer.mozilla.org/en-US/docs/Web/CSS/font-family)
+     * @param  {String} fontSize   A valid [CSS font-size](https://developer.mozilla.org/en-US/docs/Web/CSS/font-size?v=control)
+     * @return {CanvasRenderingContext2D}
+     */
+    _get2dMeasureCanvas: function(fontName, fontSize) {
+      this.__measureCanvasCache = this.__measureCanvasCache || {};
+      var fontInfo = fontSize + ' ' + fontName;
+      if (this.__measureCanvasCache[fontInfo]) return this.__measureCanvasCache[fontInfo];
```
Comment (davidrleonard) [inline px-app-nav-behavior-measure-text.es6.js]: Makes sense. Should we create a new 2d context on the fly from a single cached canvas? Or just reuse the same 2d context interface and update the fontSize/fontName on the fly if the next request has a different fontName/fontSize?

File: px-app-nav-group.html [hunk 2]
```diff
@@ -0,0 +1,486 @@
+<link rel="import" href="../polymer/polymer.html">
+<link rel="import" href="../iron-dropdown/iron-dropdown.html" />
+<link rel="import" href="../neon-animation/neon-animations.html">
```
Comment (davidrleonard) [inline px-app-nav-group.html]: Huge bummer. OK, guess I'll have to do this myself then.

File: px-app-nav-group.html [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+      },
+
+      /**
+       * A named path for this item. Used to select one of its subitems programatically,
+       * e.g. by binding to the current URL route. Path names cannot have a dot
+       * ('.') in them or they will not work.
+       */
+      path: {
+        type: String,
+        reflectToAttribute: true
```
Comment (davidrleonard) [inline px-app-nav-group.html]: 👍

File: px-app-nav-group.html [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+       * user interaction or through automatic selection.
+       *
+       * Set the `selected` attribute on one of the group's subitems to select
+       * this group.
+       */
+      selected: {
+        type: Boolean,
+        value: false,
+        notify: true,
+        reflectToAttribute: true,
```
Comment (davidrleonard) [inline px-app-nav-group.html]: This one we do need to reflect (on `path` we don't). We use `selected` to style with CSS, so need to ensure property changes through data binding trigger the attribute to update so the DOM picks up the changes.

File: px-app-nav-behavior-measure-text.es6.js (hunk 1 above)
Comment (benoitjchevalier) [inline px-app-nav-behavior-measure-text.es6.js]: I think we can reuse the cached context and update font size/name if needed.

File: px-app-nav-group.html [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+
+    openGroup() {
+      var group = this.$$('#group');
+      if (!group) return;
+
+      // Measure the size of the group's container before trying to open it,
+      // so we can set the min width of the group's dropdown
+      //
+      // Ben says this is OK.
+      var button = this.$$('#itemButton');
```
Comment (davidrleonard) [inline px-app-nav-group.html]: 👍

File: px-app-nav-behavior-measure-text.es6.js (hunk 1 above)
Comment (davidrleonard) [inline px-app-nav-behavior-measure-text.es6.js]: 👍 Makes sense. I updated it to cache the 2d context and update font info if necessary.

Comment (davidrleonard) [inline px-app-nav-behavior-measure-text.es6.js]: 19ed84f3252ee26cf97c5b6d269a4fbab89bae9e

File: sass/px-app-nav.scss [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+// Settings
+@import "px-colors-design/_settings.colors.scss";
+
+// Generic
+@import "px-box-sizing-design/_generic.box-sizing.scss";
+
+// Base
+@import "px-typography-design/_base.typography.scss";
```
Comment (Menkhaus-ge) [inline sass/px-app-nav.scss]: need to double check with Martin but I don't think we want to directly import typography and that it should only be include through theme.

File: sass/px-app-nav.scss [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  width: 100%;
+}
+
+.app-nav {
+  display: flex;
+}
+
+.app-nav__items {
+  height: $px-app-nav-height;
+  flex: 1;
```
Comment (Menkhaus-ge) [inline sass/px-app-nav.scss]: Should set to `flex: 1 1 auto`
IE interprets 'flex: 1' to be 'flex: 1 1 0px',

File: sass/px-app-nav.scss [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+@import "px-colors-design/_settings.colors.scss";
+
+// Generic
+@import "px-box-sizing-design/_generic.box-sizing.scss";
+
+// Base
+@import "px-typography-design/_base.typography.scss";
+
+// Component
+@function rem($size) {
```
Comment (Menkhaus-ge) [inline sass/px-app-nav.scss]: curious as to why you set your own rem function that is different than the one that we use in px-functions-design: https://github.com/PredixDev/px-functions-design/blob/master/_tools.functions.scss#L33

File: sass/px-app-nav.scss [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  $rem: $size / $inuit-base-font-size;
+  @return $rem * 1rem;
+}
+
+////// Theming
+$px-app-nav-height: var(--px-app-nav-height, rem(60px));
+$px-app-nav-background-color: var(--px-app-nav-background-color, $gray2);
+$px-app-nav-box-shadow: var(--px-app-nav-box-shadow, 0 rem(2px) rem(4px) 0 rgba(0,0,0,0.2));
```
Comment (Menkhaus-ge) [inline sass/px-app-nav.scss]: at some point we will need to update to use px-shadows-design

File: sass/px-app-nav-subitem.scss [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  position: relative;
+  display: flex;
+  align-items: center;
+  padding-left: $inuit-base-spacing-unit;
+  padding-right: $inuit-base-spacing-unit;
+  cursor: pointer;
+
+  height: $px-app-nav-subitem-height;
+  color: $px-app-nav-subitem-text-color;
+  background-color: var(--px-app-nav-subitem-background-color, transparent);
```
Comment (Menkhaus-ge) [inline sass/px-app-nav-subitem.scss]: This line is defined twice. Here and line 29.

File: sass/px-app-nav-subitem.scss [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+/* Subitem State: Selected */
+$px-app-nav-subitem-text-color--selected: var(--px-app-nav-subitem-text-color--selected, $white);
+$px-app-nav-subitem-background-color--selected: var(--px-app-nav-subitem-background-color--selected, $select-default);
+:host([selected]) {
+  color: $px-app-nav-subitem-text-color--selected;
+  background-color: $px-app-nav-subitem-background-color--selected;
+}
+:host([selected]):before {
+  display: none;
```
Comment (Menkhaus-ge) [inline sass/px-app-nav-subitem.scss]: I'm not sure when this code gets used.

File: sass/px-app-nav-subitem.scss [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+/* Subitem State: Collapsed+Parent Selected+NOT selected */
+:host([collapsed][parent-selected]) {
+  color: $gray15;
+  background-color: $white;
+}
+:host([collapsed][parent-selected]):before {
+  content: '';
+  display: block;
+  background-color: $white;
```
Comment (Menkhaus-ge) [inline sass/px-app-nav-subitem.scss]: background-color is defined twice

File: sass/px-app-nav-subitem.scss [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+:host([collapsed][parent-selected]:hover) {
+  background-color: $gray1;
+}
+
+/* Subitem State: Collapsed+Parent Selected+Selected */
+:host([collapsed][parent-selected][selected]) {
+  color: $white;
+  background-color: $select-default;
+}
+:host([collapsed][parent-selected][selected]:hover) {
```
Comment (Menkhaus-ge) [inline sass/px-app-nav-subitem.scss]: Should not need this unless you want the colors to be different than `:host([collapsed][parent-selected][selected]){}` or if you want it to be configurable

File: sass/px-app-nav-item.scss [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+  padding-left: $inuit-base-spacing-unit;
+  padding-right: $inuit-base-spacing-unit;
+  cursor: pointer;
+  color: $px-app-nav-item-text-color;
+  background-color: $px-app-nav-item-background-color;
+  --iron-icon-fill-color: $px-app-nav-item-icon-color;
+}
+
+/* Item State: Pressed */
+$px-app-nav-item-icon-color--pressed: var(--px-app-nav-item-icon-color--pressed, $gray15);
```
Comment (Menkhaus-ge) [inline sass/px-app-nav-item.scss]: When I press on the app nav item I don't see a color change.

File: px-app-nav-group.html (hunk 3 above)
Comment (davidrleonard) [inline px-app-nav-group.html]: Done as of 3d2d6198e22316740b9877c9ac5ce7de5c27a921

File: px-app-nav-group.html (hunk 5 above)
Comment (davidrleonard) [inline px-app-nav-group.html]: Done as of 1ce203425f0a503bcab77cbd164a7053f2cd2eec

File: px-app-nav-group.html [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+    },
+
+    _handleDropdownOpened(evt) {
+      evt.preventDefault();
+      evt.stopPropagation();
+
+      const content = Polymer.dom(this.root).querySelector('#groupcontainer');
+      const detail = {};
+      if (content) {
+        const rect = content.getBoundingClientRect()
```
Comment (davidrleonard) [inline px-app-nav-group.html]: ✅  Removed as of 3b88ee54d6a18ad7bdd76134ee372164a3faa1ae

File: px-app-nav-group.html [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+    },
+
+    openGroup() {
+      var group = this.$$('#group');
+      if (!group) return;
+
+      // Measure the size of the group's container before trying to open it,
+      // so we can set the min width of the group's dropdown
+      //
+      // Ben says this is OK.
```
Comment (davidrleonard) [inline px-app-nav-group.html]: ✅ 😉

File: px-app-nav-group.html (hunk 4 above)
Comment (davidrleonard) [inline px-app-nav-group.html]: ❌

File: sass/px-app-nav.scss (hunk 6 above)
Comment (davidrleonard) [inline sass/px-app-nav.scss]: ✅  fixed in 4f141412a88cceb90a18c310238a60e91b5d74fc

File: sass/px-app-nav.scss (hunk 8 above)
Comment (davidrleonard) [inline sass/px-app-nav.scss]: ✅  fixed in 449b7e3ea687c5bc5d4aa389f9c3d01e524b0c5c

File: sass/px-app-nav.scss (hunk 9 above)
Comment (davidrleonard) [inline sass/px-app-nav.scss]: ✅  fixed in 1fbec9d20d1121bedf89b6dfc2de6ed98303a375

File: sass/px-app-nav.scss (hunk 7 above)
Comment (davidrleonard) [inline sass/px-app-nav.scss]: ✅  fixed in 7afde503c865ba6ede830b4ebca418f99d4aa4bc

File: sass/px-app-nav-subitem.scss (hunk 10 above)
Comment (davidrleonard) [inline sass/px-app-nav-subitem.scss]: ✅  fixed in 777820a6b1df353d5709c830f7ce1599784860ee

File: sass/px-app-nav-subitem.scss (hunk 11 above)
Comment (davidrleonard) [inline sass/px-app-nav-subitem.scss]: ✅  fixed in 777820a6b1df353d5709c830f7ce1599784860ee

File: sass/px-app-nav-subitem.scss (hunk 12 above)
Comment (davidrleonard) [inline sass/px-app-nav-subitem.scss]: ✅  fixed in 777820a6b1df353d5709c830f7ce1599784860ee

File: sass/px-app-nav-subitem.scss (hunk 13 above)
Comment (davidrleonard) [inline sass/px-app-nav-subitem.scss]: ✅  fixed in 777820a6b1df353d5709c830f7ce1599784860ee

File: sass/px-app-nav-item.scss (hunk 14 above)
Comment (davidrleonard) [inline sass/px-app-nav-item.scss]: ✅  fixed in 777820a6b1df353d5709c830f7ce1599784860ee

**Pull request merged**